src/commands/synthetics/command.ts

```diff
@@ -114,53 +113,66 @@ export class RunTestCommand extends Command {
     triggers: Trigger,
     startTime: number
   ) {
-    // Sort tests to show success first then non blocking failures and finally blocking failures.
-    tests.sort(this.sortTestsByOutcome(results))
+    const getTest = (publicId: string) => tests.find((t) => t.public_id === publicId) as Test
+
+    const sortedPollResults = Object.entries(results)
+      .reduce(
+        (accResults, [publicId, pollResults]) => accResults.concat(pollResults.map((r) => [getTest(publicId), r])),
+        [] as [Test, PollResult][]
+      )
+      .sort(this.sortResultsByOutcome())
 
     // Rendering the results.
     this.reporter?.reportStart({startTime})
+
     const locationNames = triggers.locations.reduce((mapping, location) => {
       mapping[location.id] = location.display_name
 
       return mapping
     }, {} as LocationsMapping)
+
+    if (!this.config.failOnTimeout) {
+      if (!summary.timedOut) {
+        summary.timedOut = 0
+      }
+    }
+
+    if (!this.config.failOnCriticalErrors) {
+      if (!summary.criticalErrors) {
+        summary.criticalErrors = 0
+      }
+    }
+
     let hasSucceeded = true // Determine if all the tests have succeeded
-    for (const test of tests) {
-      const testResults = results[test.public_id]
-      if (!this.config.failOnTimeout) {
-        if (!summary.timedOut) {
-          summary.timedOut = 0
-        }
 
-        const hasTimeout = testResults.some((pollResult) => pollResult.result.error === ERRORS.TIMEOUT)
-        if (hasTimeout) {
-          summary.timedOut++
-        }
+    for (const [test, pollResult] of sortedPollResults) {
+      if (!this.config.failOnTimeout && pollResult.result.error === ERRORS.TIMEOUT) {
+        summary.timedOut++
       }
 
-      if (!this.config.failOnCriticalErrors) {
-        if (!summary.criticalErrors) {
-          summary.criticalErrors = 0
-        }
-        const hasCriticalErrors = testResults.some((pollResult) => isCriticalError(pollResult.result))
-        if (hasCriticalErrors) {
-          summary.criticalErrors++
-        }
+      if (!this.config.failOnCriticalErrors && isCriticalError(pollResult.result)) {
+        summary.criticalErrors++
       }
 
-      const passed = hasTestSucceeded(testResults, this.config.failOnCriticalErrors, this.config.failOnTimeout)
-      if (passed) {
+      const resultOutcome = getResultOutcome(
+        test,
+        pollResult,
+        this.config.failOnCriticalErrors,
+        this.config.failOnTimeout
+      )
+
+      if ([ResultOutcome.Passed, ResultOutcome.PassedNonBlocking].includes(resultOutcome)) {
         summary.passed++
+      } else if (resultOutcome === ResultOutcome.FailedNonBlocking) {
+        summary.failedNonBlocking++
       } else {
         summary.failed++
-        if (test.options.ci?.executionRule !== ExecutionRule.NON_BLOCKING) {
-          hasSucceeded = false
-        }
+        hasSucceeded = false
       }
 
       this.reporter?.testEnd(
```

Comment:
So now `testEnd()` would be called for every result instead of once per test when all its results are received?
If so it mimics the `testResult()` (in latest `master`), [doc](https://github.com/DataDog/datadog-ci/blob/8659032811231c6f8918c633d5f703e659963f32/src/commands/synthetics/README.md#L263) might need to be updated. 
Also it feels a bit strange to always invoke reporters with an array of a single item, if necessary we can redefine the reporter interfaces (maybe to see with @sdeprez now what make the most sense to also work with batches later since they'll probably need to be updated anyway).

Reply:
The `testResult()` hook is called whenever a _finished_ result has been polled. Here, we need to wait for all the results to be polled, because we have to sort them by outcome (passed > failed non blocking > failed) before displaying them in the report.

The `testEnd()` name now feels weird indeed, but I didn't want to change the `MainReporter` interface. As you mentioned, I would be in favour of re-designing a better interface, most likely with breaking changes 🙂